### PR TITLE
[extension/ballastextension] Deprecate memory_ballast extension

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** 

Based on user reports on open-telemetry/opentelemetry-helm-charts/issues/891 and the discussion on #8343, we can deprecate the memory ballast extension in favor of using `GOMEMLIMIT`. This PR:

- Deprecates the memory ballast extension in the README
- Removes references to the memory ballast extension on docs
- Updates k8s example to use `GOMEMLIMIT` with the same approach as in the Helm chart (80% of memory limit)
- Deprecates the memory ballast extension Go module

Once this PR is accepted, open-telemetry/opentelemetry-helm-charts/issues/891 can move ahead with enabling `useGOMEMLIMIT` by default on the Helm chart.

Other issues will be opened for opentelemetry.io, the Opentelemetry Operator and other parts of the OpenTelemetry project to remove references to this extension once the PR is merged.

No explicit timeline is given for removal of the extension.

**Link to tracking Issue:** Updates #8343

### CODE REVIEW

**Comment by mx-psi** [comment]
@dmitryax @Aneurysm9 @TylerHelmuth @JaredTan95

**Review by dloucasfx — COMMENTED** [review]
Thanks @mx-psi for pushing this and getting us closer to GOMEMLIMIT !

Not sure if this was already discussed, the ballast extension has been used in conjunction with the memory-limiter processor to control GC; 

With `GOMEMLIMIT` do we need to change the behavior of memory-limiter ? 
The memory-limiter currently provides 2 main tasks:
- signal the pipeline to drop or stop accepting new data
- call GC explicitly

With `GOMEMLIMIT` we should no longer worry about explicitly calling GC anymore, however, we might still need to keep the drop or stop accepting data? 
I also understand that @dmitryax suggested to move the memory-limiter as a receiver helper.

File: extension/ballastextension/README.md [hunk 1]
```diff
@@ -1,8 +1,12 @@
+> [!WARNING]
+> The memory ballast extension is deprecated in favor of using the `GOMEMLIMIT` environment variable.
+> This environment variable is available on any Collector built with Go 1.19 or higher. Official binary releases are built with Go 1.19 since v0.61.0. See [issue 8343](https://github.com/open-telemetry/opentelemetry-collector/issues/8343) for the deprecation timeline.
```
Comment (codeboten) [inline extension/ballastextension/README.md:3]: Should this include a recommendation of what a user should configure this based on their previous setting?

**Comment by bogdandrutu** [comment]
@mx-psi would be good to look into the memory limiter as well.

**Comment by mx-psi** [comment]
@bogdandrutu @dloucasfx There are a couple dependencies on the memory ballast in other parts of the Collector:
- Issue #8632 summarizes the work needed to get rid of the memory limiter and moving to receivers
- Internal metrics also have a dependency on the memory ballast, which is summarized on #8342

IMO, we should first
1. deprecate the memory ballast
2. update all our documentation and examples to stop mentioning the memory ballast
3. update the Helm chart, operator and Docker default configuration to use GOMEMLIMIT
4. (can happen in parallel with 2 & 3) wait for a couple of releases

After these three happen, we can think about changing the behavior of other parts of the Collector that depend on the memory ballast, including the memory limiter (if the receiver part) and the metrics. Does that make sense? Do you think there is specific parts we should do now as part of this PR?

**Comment by mx-psi** [comment]
@bogdandrutu @dloucasfx can you take another look?

**Comment by mx-psi** [comment]
Looks like we are first going to enable `GOMEMLIMIT` by default in the Helm Chart for a couple of versions before doing this, you can follow along at open-telemetry/opentelemetry-helm-charts/issues/891

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**